jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +170,419 @@ default void clear()
+        ByteBuffer to = getByteBuffer();
+
+        int pos = BufferUtil.flipToFill(to);
+        try
+        {
+            length = Math.min(length, to.remaining());
+            to.put(bytes, offset, length);
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(to, pos);
+        }
+        return length;
+    }
+
+    default void append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+    }
+
+    default boolean append(ByteBuffer bytes, Runnable releaser)
+    {
+        append(bytes);
+        if (bytes.hasRemaining())
+            return false;
+        releaser.run();
+        return true;
+    }
+
+    default void append(RetainableByteBuffer bytes)
+    {
+        bytes.writeTo(this);
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        toInfillMode.put(slice);
+    }
+
+    default boolean writeTo(ByteBuffer to)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        BufferUtil.append(to, slice);
+        return slice.remaining() == 0;
+    }
+
+    default boolean writeTo(RetainableByteBuffer to)
+    {
+        if (remaining() == 0)
+            return true;
+        ByteBuffer slice = getByteBuffer().slice();
+        to.append(slice);
+        return slice.remaining() == 0;
+    }
+
+    default void writeTo(OutputStream out) throws IOException
+    {
+        BufferUtil.writeTo(getByteBuffer(), out);
+    }
+
+    default void writeTo(Content.Sink sink, boolean last, Callback callback)
+    {
+        sink.write(last, getByteBuffer(), callback);
+    }
+
+    static RetainableByteBuffer newAggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
+    {
+        return new Aggregator(pool, direct, growBy, maxCapacity);
+    }
+
+    class Aggregator implements RetainableByteBuffer
+    {
+        private final ByteBufferPool _pool;
+        private final boolean _direct;
+        private final int _growBy;
+        private final int _maxCapacity;
+        private RetainableByteBuffer _buffer;
+
+        private Aggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
+        {
+            if (growBy < 0)
+                throw new IllegalArgumentException("growBy must be > 0");
+            if (maxCapacity > 0 && growBy > maxCapacity)
+                throw new IllegalArgumentException("growBy must be < maxSize");
+
+            _pool = pool == null ? new ByteBufferPool.NonPooling() : pool;
+            _direct = direct;
+            _growBy = growBy;
+            _maxCapacity = maxCapacity <= 0 ? Integer.MAX_VALUE : maxCapacity;
+            _buffer = _pool.acquire(_growBy, _direct);
+        }
+
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            return _buffer.getByteBuffer();
+        }
+
+        @Override
+        public int capacity()
+        {
+            return Math.max(_buffer.capacity(), _maxCapacity);
```

Comment:
I think this should be `Math.min()` as I don't think we want to overshoot `_maxCapacity`, even if the buffer is bigger than it.

Reply:
The style of interaction with RetainableByteBuffer has always been that you can bypass the API by getting the buffer directly and manipulating it.  So any capacity limit we put in this method can by bypassed anyway.
So I think it is better to embrace the fact that the pool might give us a bigger buffer than was requested and to use that extra capacity if already allocated.  I'll javadoc that.

Reply:
Alternately we could have an extra constructor that takes an additional `strictCapacity` boolean, that if true would always slice a buffer acquired from the pool to be exactly the desired capacity.

Or should we just have a variant of the pool that does that?